### DISCUSSION TITLE
Unhandeled CastError in dataloader

### DISCUSSION THREAD

**Discussion by chladog** [payloadcms/payload#911] [Feature Requests & Ideas]
I encoutered following error already with #884 which since then was fixed.
Now the topic is regarding an error handling of this error. Sometimes I get this error, probably because of old invalid type data in my mongodb. Ofc invalid data shouldn't get stored in DB in the first place, but it happens - in this particular case it was a saved JSON in the relationship field where only ID should be stored.

The problem is that this error currently 1) kills the server process 2) therefore breaks whole GQL request so it's not possible to generate at least partial view with valid data and displaying an error to user.

```JS
const castError = new CastError();
                    ^
CastError: Cast to ObjectId failed for value "[object Object]" (type string) at path "_id" for model "media"
    at model.Query.exec (\node_modules\mongoose\lib\query.js:4777:21)
    at Function.paginate (\node_modules\mongoose-paginate-v2\dist\index.js:197:26)
    at find (\node_modules\payload\src\collections\operations\find.ts:146:37)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async \node_modules\payload\src\collections\dataloader.ts:66:20
    at async Promise.all (index 3)
    at async DataLoader._batchLoadFn (\node_modules\payload\src\collections\dataloader.ts:97:3) {
  messageFormat: undefined,
  stringValue: '"[object Object]"',
  kind: 'ObjectId',
  value: '[object Object]',
  path: '_id',
  reason: BSONTypeError: Argument passed in must be a string of 12 bytes or a string of 24 hex characters or an integer
      at new BSONTypeError (\node_modules\bson\src\error.ts:16:5)
      at new ObjectId (\node_modules\bson\src\objectid.ts:88:15)
      at castObjectId (\node_modules\mongoose\lib\cast\objectid.js:25:12)
      at ObjectId.cast (\node_modules\mongoose\lib\schema\objectid.js:246:12)
      at ObjectId.SchemaType.applySetters (\node_modules\mongoose\lib\schematype.js:1189:12)
      at ObjectId.SchemaType._castForQuery (\node_modules\mongoose\lib\schematype.js:1623:15)
      at ObjectId.SchemaType.castForQuery (\node_modules\mongoose\lib\schematype.js:1613:15)
      at \node_modules\mongoose\lib\schematype.js:1555:18
      at Array.map (<anonymous>)
      at ObjectId.handle$in (\node_modules\mongoose\lib\schematype.js:1551:14),
  valueType: 'string'
```

Could this be handled the way that as much of valid data is returned back along with error message reporting the field of the problem so it doesn't break the app UX?

**Comment by chladog** [comment]
I just found out that this error also happens when querying object that has relationship field reference to an object that doesn't exist anymore and it tries to query it (when depth > 0)
_(e.g. media has a field "**owner**" related to ```users``` with ID of **userA**, then you delete **userA** without updating media field owner and query media with depth > 0)_

This should be moved to Issues.

**Reply by denolfe** [reply]
Issue created #923

**Reply by jmikrut** [reply]
Hey @chladog — I just tried to replicate this exact issue and can't seem to reproduce. I deleted a document, and then queried a parent document that still had a reference to the deleted child, and REST simply returns the ID of the now-deleted doc, and GraphQL returns `null` as expected.

There are a few different items being discussed in this discussion, and if we can reproduce, this would certainly warrant an issue. But I can't reproduce. Can you help me figure out how to replicate?

As for better handling errors pertaining to improper data—I'll tackle that separately as a feature enhancement shortly. 👍

**Reply by chladog** [reply]
Hi @jmikrut 
Great. Yes I realize this is not good bug report :-D I'll try to replicate in fresh install and share with reproduction steps.

**Comment by chladog** [comment]
This also happens randomly on existing documents with relationship fields. I suspect that relationship field JSONs are being mixed with relationship fields IDs in dataloader.

I have following local API query in custom GQL resolver:
```JS
            const result = await payload.find({
              collection: "pages",
              where,
              user: context.req.user,
              locale: context.req.locale,
              fallbackLocale: context.req.fallbackLocale,
              depth: 1,
            });
```

Stored document looks like this:
![image](https://user-images.githubusercontent.com/15977485/184368411-49d13ed9-62a7-4fcd-a50a-cd15eefcb2a4.png)

Error I'm getting:
```Cast to ObjectId failed for value \"[object Object]\" (type string) at path \"_id\" for model \"users\"```

When I set depth: 0 I would expect id of a user, however I get user's JSON.
